### PR TITLE
Calculator: Support repeating, chaining operations and other improvements

### PR DESCRIPTION
**Pull request by karolba**
Make the calculator behave a bit more like a physical desktop calculator:

- Fix key animations when using the keyboard (ec3ec22)
- Add support for chaining binary operations, like "1+2+3" (which would previously evaluate as 5)  
GIF of 1+2+3
- Add support for hitting "=" repeatedly to repeat last addition/subtraction/multiplication/division  
GIF of repeatedly hitting "="
- Make the percent key calculate a percentage of the left side o when used in an addition/subtraction/multiplication/division  
GIF of calculating 25% of 30
- Make it possible to change our mind when selecting addition/subtraction/multiplication/division  
GIF of "2+-*/3"

:^)

### CODE REVIEW

File: Userland/Applications/Calculator/CalculatorWidget.cpp [hunk 1]
```diff
@@ -159,7 +159,8 @@ void CalculatorWidget::keydown_event(GUI::KeyEvent& event)
         m_keypad.set_value(m_calculator.finish_operation(m_keypad.value()));
         mimic_pressed_button(m_equals_button);
     } else if (event.code_point() >= '0' && event.code_point() <= '9') {
-        auto const digit = m_keypad.type_digit(event.code_point() - '0');
+        auto const digit = event.code_point() - '0';
+        m_keypad.type_digit(digit);
```
Comment (LucasChollet) [inline Userland/Applications/Calculator/CalculatorWidget.cpp:166]: Can you change the return type of `type_digit()` to be `void`?
I think that it was its only usage in the codebase.

File: Userland/Applications/Calculator/Calculator.cpp [hunk 2]
```diff
@@ -22,78 +26,117 @@ Crypto::BigFraction Calculator::begin_operation(Operation operation, Crypto::Big
...
 }
 
-Crypto::BigFraction Calculator::finish_operation(Crypto::BigFraction argument)
+Optional<Crypto::BigFraction> Calculator::operation_without_argument(Operation operation)
+{
+    bool in_binary_operation = m_binary_operation_in_progress != Operation::None;
+    bool entering_new_binary_operation = operation_is_binary(operation);
+    bool previous_operation_was_binary = operation_is_binary(m_previous_operation);
+
+    if (in_binary_operation && entering_new_binary_operation) {
```
Comment (LucasChollet) [inline Userland/Applications/Calculator/Calculator.cpp:110]: Small nitpick on the style.

1. We usually don't put brackets around single line `if` and `else`
2. We don't use `else` after a `return`

File: Userland/Applications/Calculator/Calculator.h [hunk 3]
```diff
@@ -36,20 +37,30 @@ class Calculator final {
...
+    Crypto::BigFraction m_current_value {};
+
+    Operation m_binary_operation_in_progress { Operation::None };
+    Crypto::BigFraction m_binary_operation_saved_left_side {};
+
+    Operation m_previous_operation { Operation::None };
+    Crypto::BigFraction m_previous_binary_operation_right_side {};
     bool m_has_error { false };
+
+    Crypto::BigFraction finish_binary_operation(Crypto::BigFraction left_side, Operation operation, Crypto::BigFraction right_side);
```
Comment (LucasChollet) [inline Userland/Applications/Calculator/Calculator.h:65]: `Crypto::BigFraction` can be huge, let's avoid a copy and take a `const&` here

**Review by LucasChollet — COMMENTED** [review]
Nice to see the Calculator gets some love :^)

File: Userland/Applications/Calculator/CalculatorWidget.cpp (hunk 1 above)
Comment (karolba) [reply Userland/Applications/Calculator/CalculatorWidget.cpp:166]: Done

File: Userland/Applications/Calculator/Calculator.h (hunk 3 above)
Comment (karolba) [reply Userland/Applications/Calculator/Calculator.h]: Done

File: Userland/Applications/Calculator/Calculator.cpp (hunk 2 above)
Comment (karolba) [reply Userland/Applications/Calculator/Calculator.cpp:110]: Done

**Pull request merged**